[Core] StateHash Interfaces & Diagrams (#252)
pokt-network/pocket: shared/docs/PROTOCOL_STATE_HASH.md

# State Hash <!-- omit in toc -->

This document describes the cross-module communication using the interfaces in [../shared/modules](../shared/modules) to compute a new state hash. See module specific documentation & implementation details inside each module respectively.

- [Context Management](#context-management)
- [Block Application](#block-application)

_NOTE: The diagrams below use some [Hotstuff specific](https://arxiv.org/abs/1803.05069) terminology as described in the [HotPOKT Consensus Specification](https://github.com/pokt-network/pocket-network-protocol/tree/main/consensus) but can be adapted to other BFT protocols as well._

<!-- See if there's an answer in this question to add links to notes: https://stackoverflow.com/questions/74103729/adding-hyperlinks-to-notes-in-mermaid-sequence-diagrams -->

## Context Management

> The `Utility` and `Persistence` modules maintain a context (i.e. an ephemeral states) driven by the `Consensus` module that can be released & reverted as a result of various (e.g. lack of Validator consensus) before the state is committed and persisted to disk (i.e. the block is finalized).

Reviewer: ```suggestion
The `Utility` and `Persistence` modules maintain a context (i.e. an ephemeral states) driven by the `Consensus` module that can be `released & reverted` (i.e. the block is invalid / no Validator Consensus reached) or can be `committed & persisted` to disk (i.e. the block is finalized).
```

On every round of every height:

1. The `Consensus` module handles a `NEWROUND` message
2. A new `UtilityContext` is initialized at the current height
3. A new `PersistenceRWContext` is initialized at the current height
4. The [Block Application](#block-application) flow commences

```mermaid
sequenceDiagram
    title Steps 1-4
    participant B as Bus
    participant C as Consensus
    participant U as Utility
    participant P as Persistence

    %% Handle New Message

>     B-->>C: HandleMessage(msg)

Reviewer: Let's make it explicit that `msg` is `newRoundMsg`

Ditto with `DECIDE`

    critical NewRound
        %% Create Contexts
        C->>+U: NewContext(height)
        U->>+P: NewRWContext(height)
        P->>-U: PersistenceContext
        U->>U: store context<br>locally
        activate U
        deactivate U
        U->>-C: UtilityContext
        C->>C: store context<br>locally
        activate C
        deactivate C

        %% Apply Block

>         Note over C, P: See 'Block Application'

Reviewer: Omit `See` to be consistent with the chart below's reference to `StoreBlock`

    end
```

> 5. The **HotPOKT lifecycle** takes place so Validators achieve consensus (i.e. steps `PRECOMMIT` and `COMMIT`)

Reviewer: 5 is orphaned between two charts. This is confusing. 

It also implies that `PRECOMMIT` and `COMMIT` are the entirety of the `HotPOKT lifecycle`

Please capitalize `Lifecycle` if you're going to bold it

6. The `Consensus` module handle the `DECIDE` message

> 7. The final `quorumCertificate` is propagated to the `UtilityContext` on Commit

Reviewer: Why do we need to use the Utility Module to `passthrough` the `QC` to the `persistenceModule`?

I.E. can 7&8 be combined?

8. The final `quorumCertificate` is propagated to the `PersistenceContext` on Commit

> 9. The persistence module's internal implementation for [Store Block](../../persistence/docs/PROTOCOL_STATE_HASH.md) must execute.

Reviewer: **Okay so this is the big one @Olshansk**

*DISCUSS*

We need the `AppHash` during `ApplyBlock` to come to `Consensus` on whether or not the `ProposalBlock` is Valid. So something is wrong here in the ordering as Step 9 (PROTOCOL_STATE_HASH) should have been executed prior to Step 5...

Maybe I'm misunderstanding what you're implying here by `StateHash` but either way - we need to discuss this

Reviewer: I notice that below:

```markdown
The persistence module's internal implementation for [Updating a State hash] must be triggered
```
You have the lifecyle trigger properly. So I'm not sure why we are `double executing` the StateHash execution... Sorry if I'm misunderstanding what you're saying

Author: We're on the same page. I updated some links and terminology to reflect the confusion. 

Should be clearer now.

10. Both the `UtilityContext` and `PersistenceContext` are released

```mermaid
sequenceDiagram
    title Steps 6-10
    participant B as Bus
    participant C as Consensus
    participant U as Utility
    participant P as Persistence

    %% Handle New Message
    B-->>C: HandleMessage(msg)

    critical Decide
        %% Commit Context
        C->>+U: Context.Commit(quorumCert)
        U->>+P: Context.Commit(quorumCert)
        P->>P: Internal Implementation
        Note over P: Store Block
        P->>-U: err_code
        U->>C: err_code
        deactivate U

        %% Release Context
        C->>+U: Context.Release()
        U->>+P: Context.Release()
        P->>-U: err_code
        U->>-C: err_code
    end

```

## Block Application

When applying the block during the `NEWROUND` message shown above, the majority of the flow is similar between the _leader_ and the _replica_ with one of the major differences being a call to the `Utility` module as seen below.

- `ApplyBlock` - Uses the existing set of transactions to validate & propose
- `CreateAndApplyProposalBlock` - Reaps the mempool for a new set of transaction to validate and propose

```mermaid

> graph TD

Reviewer: Cool chart! Very readable!

    B[Should I prepare a new block?] --> |Wait for 2/3+ NEWROUND messages| C

    C[Am I the leader?] --> |Yes| D
    C[Am I the leader?] --> |No| Z

>     D[Did I get any prepareQCs?] --> |Find highest valid PrepareQC| E

Reviewer: Question:

Does `HighPrepareQC` imply a `LockedBlock`?

Reviewer: If a proposer dies before they submit the block it's a `locked_qc`. So a `locked_qc` can be a `high_prepare_qc`?

Reviewer: If not, let's include it in your chart

Author: > So a `locked_qc` can be a `high_prepare_qc`

Yes.  A `lockedQC` may be either `nil` or `prepareQC`, and since we aggregate `prepareQCs`, it may be a `highPrepareQC`. It looks something like this:

| QC        | NewRound | Prepare | Precommit | Commit | Decide    |
| --------- | -------- | ------- | --------- | ------ | --------- |
| prepareQC | nil      | nil     | QC        | QC     | QC -> nil |
| lockedQC  | nil      | nil     | nil       | QC     | QC -> nil |

--- 

I updated the diagram to make it more explicit but not sure if answered the question directly. Lmk

```mermaid
sequenceDiagram
    title Steps 1-3
    participant C as Consensus
    participant U as Utility
    participant P as Persistence

        %% Retrieve the persistence context
        C->>+U: GetPersistenceContext()
        U->>-C: PersistenceContext

        %% Update the proposal in the persistence context
        C->>+P: SetProposalBlock
        P->>-C: err_code

        %% Apply the block to the local proposal state
        C->>+U: ApplyBlock / CreateAndApplyProposalBlock
        U->>-C: err_code
```

4. Loop over all transactions proposed
5. Check if the transaction has already been applied to the local state
6. Perform the CRUD operation(s) corresponding to each transaction
7. The persistence module's internal implementation for [Updating a State hash](../../persistence/docs/PROTOCOL_STATE_HASH.md) must be triggered
8. Validate that the local state hash computed is the same as that proposed

```mermaid
sequenceDiagram
    title Steps 4-8
    participant C as Consensus
    participant U as Utility
    participant P as Persistence

    loop for each tx in txs
        U->>+P: TransactionExists(txHash)
        P->>-U: true | false

>         opt if tx is not indexed

Reviewer: I'm not sure the `opt` is strong enough to show `loop termination` if the `TransactionExists()`

Because ofc if the `TransactionExists()` then the `Block is Invalid` and we can stop `ApplyBlock`

Author: I decided to simply change the return value to `false (does not exist)` and remove `opt` altogether.

It only shows the happy path but much clearer IMO.

            loop for each operation in tx
                U->>+P: Get*/Set*/Update*/Insert*
                P->>-U: err_code
                U->>U: Validation logic
                activate U
                deactivate U
            end
        end
    end

>     U->>+P: UpdateAppHash()

Reviewer: TODO: Consolidate `AppHash` and `StateHash`

On every round of every height:

1. The `Consensus` module handles a `NEWROUND` message
2. A new `UtilityContext` is initialized at the current height
3. A new `PersistenceRWContext` is initialized at the current height
4. The [Block Application](#block-application) flow commences

```mermaid
sequenceDiagram
    title Steps 1-4
    participant B as Bus
    participant C as Consensus
    participant U as Utility
    participant P as Persistence

    %% Handle New Message
    B-->>C: HandleMessage(NEWROUND)

    %% NEWROUND

    activate C
    %% Create Contexts
    C->>+U: NewContext(height)
    U->>+P: NewRWContext(height)
    P->>-U: PersistenceContext
    U->>U: store context<br>locally
    activate U
    deactivate U
    U->>-C: UtilityContext
    C->>C: store context<br>locally
    deactivate C

    %% Apply Block
    Note over C, P: 'Block Application'
```

---

_The **Proposer** drives the **Validators** to agreement via the **Consensus Lifecycle** (i.e. HotPOKT)_

---

5. The `Consensus` module handles the `DECIDE` message

> 6. The final `quorumCertificate` is propagated to the `UtilityContext` & `PersistenceContext` on `Commit`

Reviewer 2: In the issue description and also here, you talk about the "final quorumCertificate" but this exact same language is not used in the [Specification](https://github.com/pokt-network/pocket-network-protocol/tree/main/consensus). Emphasis on the word **final**. By intuition, I assume we are talking about the `CommitQC`. Am I correct?

If I am correct, do you reckon it would be worth it changing the wording accordingly here?

Author: Great point! Updated
